### PR TITLE
[4.14] OCPBUGS-32580: allow override of NewVolumeManagerReconstruction

### PR DESCRIPTION
**Pull request by dobsonj**
https://issues.redhat.com/browse/OCPBUGS-32580

This provides a way to set the NewVolumeManagerReconstruction feature gate in a way that still allows the cluster to be upgradable. This is only needed on 4.14, since the feature gate defaults to true on 4.15 and later.

To test, you can log into a 4.14 node, install the following test rpm, and set the environment variable in `/etc/kubernetes/kubelet-env`:

```
$ oc debug node/ip-10-0-92-74.us-east-2.compute.internal
Starting pod/ip-10-0-92-74us-east-2computeinternal-debug-nkwpd ...
To use host binaries, run `chroot /host`
Pod IP: 10.0.92.74
If you don't see a command prompt, try pressing enter.
sh-4.4# chroot /host

sh-5.1# rpm-ostree override replace https://files.dobsonj.com/openshift-hyperkube-1.27.13-1.0.401bb48.x86_64.rpm

sh-5.1# vi /etc/kubernetes/kubelet-env
sh-5.1# cat /etc/kubernetes/kubelet-env
OCP_4_14_SUPPORT_EXCEPTION_ENABLE_NEW_VOLUME_MANAGER_RECONSTRUCTION=true
sh-5.1# systemctl reboot
```

If it was successful, kubelet should have a log message showing that the environment variable took effect, and the NewVolumeManagerReconstruction feature gate should be true.

```
sh-5.1# journalctl -u kubelet | grep -m1 OCP_4_14_SUPPORT_EXCEPTION_ENABLE_NEW_VOLUME_MANAGER_RECONSTRUCTION
Apr 24 20:24:18 ip-10-0-92-74 kubenswrapper[2052]: I0424 20:24:18.141491    2052 patch_kube_features.go:19] Environment variable OCP_4_14_SUPPORT_EXCEPTION_ENABLE_NEW_VOLUME_MANAGER_RECONSTRUCTION is set, setting feature gate NewVolumeManagerReconstruction to true

sh-5.1# journalctl -u kubelet | grep -m1 'feature gates: .*NewVolumeManagerReconstruction'
Apr 24 20:24:18 ip-10-0-92-74 kubenswrapper[2052]: I0424 20:24:18.198774    2052 feature_gate.go:250] feature gates: &{map[AdmissionWebhookMatchConditions:false CloudDualStackNodeIPs:true DynamicResourceAllocation:false EventedPLEG:false MaxUnavailableStatefulSet:false NewVolumeManagerReconstruction:true NodeSwap:false RetroactiveDefaultStorageClass:false ValidatingAdmissionPolicy:false]}
```

/cc @openshift/storage @jsafrane @rphillips

### CODE REVIEW

File: pkg/features/patch_kube_features.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	if enabled {
+		klog.Infof("Environment variable %s is set, setting feature gate %s to true", overrideVariable, string(NewVolumeManagerReconstruction))
+		fg := map[string]bool{string(NewVolumeManagerReconstruction): true}
+		runtime.Must(utilfeature.DefaultMutableFeatureGate.SetFromMap(fg))
+	}
+}
+
+func init() {
+	OpenShiftNewVolumeManagerReconstructionOverride()
+}
```
Comment (dobsonj) [inline pkg/features/patch_kube_features.go:27]: In case anyone else has the same question I had: What happens when two files in a single package both have their own `init()` routine?

The files are processed alphabetically. First the init() in [kube_features.go](https://github.com/openshift/kubernetes/blob/401bb488d6f5e704de86bbc28fcdd907a130be70/pkg/features/kube_features.go#L921-L923) is called, and then the init() here in `patch_kube_features.go`. They are always called exactly once, in serial, in order.

In other words, `OpenShiftNewVolumeManagerReconstructionOverride()` should always be called after the default feature gates are added.

**Comment by duanwei33** [comment]
We can reproduce it in the latest 4.14 payload. (See reproducer in [OCPBUGS-32580](https://issues.redhat.com/browse/OCPBUGS-32580))
With the open PR + setting the environment variable in /etc/kubernetes/kubelet-env in the node, the issue is not reproduced. Although we need additional test like upgrade, but the fix looks good.

File: pkg/features/patch_kube_features.go (hunk 1 above)
Comment (jsafrane) [reply pkg/features/patch_kube_features.go:27]: It would be nice to have this information in the code.

**Comment by jsafrane** [comment]
/retest-require
this PR does not touch GCP code.

**Comment by deads2k** [comment]
@rphillips on your head be it.

/approve
